File: src/Ogooreck.Sample.BusinessLogic.Tests/Deciders/BankAccount.cs
namespace Ogooreck.Sample.BusinessLogic.Deciders;

public record BankAccountOpened(
    Guid BankAccountId,
    string AccountNumber,
    Guid ClientId,
    string CurrencyISOCode,
    DateTimeOffset CreatedAt,
    long Version
);

public record DepositRecorded(
    Guid BankAccountId,
    decimal Amount,
    Guid CashierId,
    DateTimeOffset RecordedAt,
    long Version
);

public record CashWithdrawnFromATM(
    Guid BankAccountId,
    decimal Amount,
    Guid ATMId,
    DateTimeOffset RecordedAt,
    long Version
);

public record BankAccountClosed(Guid BankAccountId,
    string commandReason,
    DateTimeOffset ClosedAt,
    long Version);

public enum BankAccountStatus
{
    Opened,
    Closed
}

public record BankAccount(
    Guid Id,
    BankAccountStatus Status,
    decimal Balance,
    long Version = 0
)
{
    public static BankAccount Evolve(BankAccount bankAccount, object @event)
    {
        return @event switch
        {
            BankAccountOpened bankAccountCreated =>
                Create(bankAccountCreated),
            DepositRecorded depositRecorded =>
                bankAccount.Apply(depositRecorded),
            CashWithdrawnFromATM cashWithdrawnFromATM =>
                bankAccount.Apply(cashWithdrawnFromATM),
            BankAccountClosed bankAccountClosed =>
                bankAccount.Apply(bankAccountClosed),
            _ => bankAccount
        };
    }

    private static BankAccount Create(BankAccountOpened @event) =>
        new BankAccount(
            @event.BankAccountId,
            BankAccountStatus.Opened,
            0,
            @event.Version
        );

    private BankAccount Apply(DepositRecorded @event) =>
        this with
        {
            Balance = Balance + @event.Amount,
            Version = @event.Version
        };

    private BankAccount Apply(CashWithdrawnFromATM @event) =>
        this with
        {
            Balance = Balance - @event.Amount,
            Version = @event.Version,
        };

    private BankAccount Apply(BankAccountClosed @event) =>
        this with
        {
            Status = BankAccountStatus.Closed,
            Version = @event.Version
        };
}

File: src/Ogooreck.Sample.BusinessLogic.FSharp.Tests/Deciders/BankAccount.fs
module Deciders.BankAccount

open System
open BankAccountPrimitives

type BankAccountOpened =
    { BankAccountId: AccountId
      AccountNumber: AccountNumber
      ClientId: ClientId
      CurrencyIsoCode: CurrencyIsoCode
      CreatedAt: DateTimeOffset
      Version: int64 }

type DepositRecorded =
    { BankAccountId: AccountId
      Amount: decimal
      CashierId: CashierId
      RecordedAt: DateTimeOffset
      Version: int64 }

type CashWithdrawnFromATM =
    { BankAccountId: AccountId
      Amount: decimal
      AtmId: AtmId
      RecordedAt: DateTimeOffset
      Version: int64 }

type BankAccountClosed =
    { BankAccountId: AccountId
      Reason: string
      ClosedAt: DateTimeOffset
      Version: int64 }

type Event =
    | BankAccountOpened of BankAccountOpened
    | DepositRecorded of DepositRecorded
    | CashWithdrawnFromAtm of CashWithdrawnFromATM
    | BankAccountClosed of BankAccountClosed

type BankAccount =
    | Initial
    | Open of
        {| Id: AccountId
           Balance: decimal
           Version: int64 |}
    | Closed of {| Id: AccountId; Version: int64 |}

let evolve (state: BankAccount) (event: Event) : BankAccount =
    match state, event with
    | Initial _, BankAccountOpened opened ->
        Open
            {| Id = opened.BankAccountId
               Balance = 0M
               Version = 1L |}
    | Open openAccount, DepositRecorded depositRecorded ->
        Open
            {| openAccount with
                Balance = openAccount.Balance + depositRecorded.Amount
                Version = depositRecorded.Version |}
    | Open openAccount, CashWithdrawnFromAtm cashWithdrawnFromAtm ->
        Open
            {| openAccount with
                Balance = openAccount.Balance - cashWithdrawnFromAtm.Amount
                Version = cashWithdrawnFromAtm.Version |}
    | Open openAccount, BankAccountClosed bankAccountClosed ->
        Closed
            {| Id = openAccount.Id
               Version = bankAccountClosed.Version |}
    | _ -> state
